File: docker/golang/01-prime-sieve/main.go
package main

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"time"
)

// ── Response ────────────────────────────────────────────────────────────────
type SieveResponse struct {
	Runtime   string `json:"runtime"`
	Limit     int    `json:"limit"`
	Count     int    `json:"count"`
	Primes    []int  `json:"primes,omitempty"`
	ElapsedUs int64  `json:"elapsed_us"`
}

// ── Algorithm ───────────────────────────────────────────────────────────────
// sieveOfEratosthenes returns every prime p where 2 ≤ p ≤ limit.
func sieveOfEratosthenes(limit int) []int {
	if limit < 2 {
		return nil
	}

	composite := make([]bool, limit+1)
	for i := 2; i*i <= limit; i++ {
		if !composite[i] {
			for j := i * i; j <= limit; j += i {
				composite[j] = true
			}
		}
	}

	primes := make([]int, 0, limit/10) // rough pre-allocation
	for i := 2; i <= limit; i++ {
		if !composite[i] {
			primes = append(primes, i)
		}
	}
	return primes
}

// ── Handlers ────────────────────────────────────────────────────────────────
func sieveHandler(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()

	limit := 10_000
	if raw := q.Get("limit"); raw != "" {
		if v, err := strconv.Atoi(raw); err == nil && v > 0 {
			if v > 10_000_000 {
				v = 10_000_000
			}
			limit = v
		}
	}
	noList := q.Get("no_list") == "true" || q.Get("no_list") == "1"

	start := time.Now()
	primes := sieveOfEratosthenes(limit)
	elapsed := time.Since(start).Microseconds()

	slog.Info("sieve completed",
		"limit", limit,
		"count", len(primes),
		"elapsed_us", elapsed,
	)

	resp := SieveResponse{
		Runtime:   "golang-docker",
		Limit:     limit,
		Count:     len(primes),
		ElapsedUs: elapsed,
	}
	if !noList {
		resp.Primes = primes
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(resp); err != nil {
		slog.Error("encode error", "err", err)
	}
}

func healthHandler(w http.ResponseWriter, _ *http.Request) {
	w.WriteHeader(http.StatusOK)
}

// ── Entry point ──────────────────────────────────────────────────────────────
func main() {
	logger := slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
		Level: slog.LevelInfo,
	}))
	slog.SetDefault(logger)

	mux := http.NewServeMux()
	mux.HandleFunc("GET /sieve", sieveHandler)
	mux.HandleFunc("GET /health", healthHandler)

	addr := ":8080"
	slog.Info("server starting", "addr", addr)

	if err := http.ListenAndServe(addr, mux); err != nil {
		slog.Error("server error", "err", err)
		os.Exit(1)
	}
}

File: docker/rust/01-prime-sieve/src/main.rs
use axum::{
    extract::Query,
    http::StatusCode,
    response::Json,
    routing::get,
    Router,
};
use serde::{Deserialize, Serialize};
use std::time::Instant;
use tracing::info;

// ── Query params ────────────────────────────────────────────────────────────
#[derive(Deserialize)]
struct SieveParams {
    /// Upper bound (inclusive). Capped at 10 000 000.
    limit: Option<u64>,
    /// When non-zero, the `primes` array is omitted from the response
    /// to avoid serialisation overhead during throughput benchmarks.
    /// Accepts "0" or "1" (serde_urlencoded rejects bool "true"/"false" strings).
    #[serde(default)]
    no_list: Option<u8>,
}

// ── Response ────────────────────────────────────────────────────────────────
#[derive(Serialize)]
struct SieveResponse {
    runtime:    &'static str,
    limit:      u64,
    count:      usize,
    #[serde(skip_serializing_if = "Option::is_none")]
    primes:     Option<Vec<u64>>,
    elapsed_us: u128,
}

// ── Algorithm ───────────────────────────────────────────────────────────────
/// Classic Sieve of Eratosthenes.
/// Returns every prime p where 2 ≤ p ≤ limit.
fn sieve_of_eratosthenes(limit: u64) -> Vec<u64> {
    if limit < 2 {
        return vec![];
    }
    let n = (limit + 1) as usize;
    let mut is_prime = vec![true; n];
    is_prime[0] = false;
    is_prime[1] = false;

    let mut i = 2usize;
    while i * i <= limit as usize {
        if is_prime[i] {
            let mut j = i * i;
            while j < n {
                is_prime[j] = false;
                j += i;
            }
        }
        i += 1;
    }

    is_prime
        .iter()
        .enumerate()
        .filter_map(|(idx, &p)| if p { Some(idx as u64) } else { None })
        .collect()
}

// ── Handlers ────────────────────────────────────────────────────────────────
async fn sieve_handler(Query(params): Query<SieveParams>) -> Json<SieveResponse> {
    let limit = params.limit.unwrap_or(10_000).min(10_000_000);

    let start  = Instant::now();
    let primes = sieve_of_eratosthenes(limit);
    let elapsed_us = start.elapsed().as_micros();

    let count = primes.len();
    info!(limit, count, elapsed_us, "sieve completed");

    Json(SieveResponse {
        runtime:    "rust-docker",
        limit,
        count,
        primes:     (params.no_list.unwrap_or(0) == 0).then_some(primes),
        elapsed_us,
    })
}

async fn health_handler() -> StatusCode {
    StatusCode::OK
}

// ── Entry point ──────────────────────────────────────────────────────────────
#[tokio::main]
async fn main() {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::from_default_env()
                .add_directive("prime_sieve=info".parse().unwrap()),
        )
        .json()
        .init();

    let app = Router::new()
        .route("/sieve",  get(sieve_handler))
        .route("/health", get(health_handler));

    let addr = "0.0.0.0:8080";
    info!("listening on {addr}");
    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
